tests/cpp/test_multidevice_host_ir.cpp

```diff
@@ -468,6 +468,77 @@ TEST_F(OverlapDistributedMatmulTest, AG_linear) {
   EXPECT_TRUE(torch::allclose(out_ref, out_at, 1e-1, 1e-1));
 }
 
+TEST_F(MultiDeviceTest, ShareIpcMemHandles) {
+  static constexpr int kTensorSize = 4;
+  static constexpr int kNumRepetitions = 10;
+
+  if (communicator_->size() < 2 || torch::cuda::device_count() < 2) {
+    GTEST_SKIP() << "This test needs at least 2 GPUs and 2 ranks.";
+  }
+
+  const DeviceIdxType my_rank = communicator_->deviceId();
+  const DeviceIdxType size = communicator_->size();
+  const DeviceIdxType send_peer = (my_rank + 1) % size;
+  const DeviceIdxType recv_peer = (size + my_rank - 1) % size;
+
+  auto container = std::make_unique<hir::HostIrContainer>();
+  FusionGuard fg(container.get());
+
+  auto* send_tv = makeContigTensor(1, DataType::Int32);
+  auto* recv_tv = makeContigTensor(1, DataType::Int32);
+
+  auto send = IrBuilder::create<P2PCommunication>(
+      P2PCommunicationType::SEND,
+      send_tv,
+      IrBuilder::create<Val>(send_peer),
+      CommunicatorBackend::kNccl);
+  auto recv = IrBuilder::create<P2PCommunication>(
+      P2PCommunicationType::RECV,
+      recv_tv,
+      IrBuilder::create<Val>(recv_peer),
+      CommunicatorBackend::kNccl);
+  std::vector<P2PCommunication*> grouped_communications = {send, recv};
+
+  ExpressionEvaluator expr_evaluator;
+  IpcHandleCache ipc_handle_cache(expr_evaluator);
+
+  auto options =
+      at::TensorOptions().dtype(at::kInt).device(communicator_->device());
+  auto generate_tensor = [options](int repetition, int rank) {
+    return at::arange(kTensorSize, options) + (repetition + 1) * 10 +
+        100 * rank;
+  };
+  at::Tensor recv_tensor = at::empty({kTensorSize}, options);
+  at::Tensor send_tensor = at::empty({kTensorSize}, options);
+
+  expr_evaluator.bind(send_tv, send_tensor);
+  expr_evaluator.bind(recv_tv, recv_tensor);
+
+  for (auto repetition : c10::irange(kNumRepetitions)) {
+    // all ranks set `send_tensor`
+    send_tensor.copy_(generate_tensor(repetition, my_rank));
+    torch::cuda::synchronize();
+    communicator_->barrier();
+
+    // Exchange IpcHandle on the first iteration
+    ipc_handle_cache.exchangeHandles(grouped_communications);
+
+    // RDMA put-zcopy
```

Comment:
What does this mean?

Reply:
RDMA = Remote Direct Memory Access
put-zcopy is the p2p protocol.
"zcopy" = zero copy, no staging of the data into internal buffers
"put", as opposed to "get" means the src writes the data to the dst

I mentioned this in my presentation